File: adt2unit/count_order.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

void count_order (uint8_t *entries)
{
  uint8_t index = 0, index2 = 0;

  DBG_ENTER ("count_order");

  do
  {
    if (songdata.pattern_order[index] != 0x80)
    {
      if (songdata.pattern_order[index] > 0x80)
      {
        if ((songdata.pattern_order[index] - 0x80) != index2)
        {
          index = songdata.pattern_order[index] - 0x80;
          index2 = index;
        }
        else
          break;
      }
    }
    else
      break;

    if (index < 0x80)
      index++;
  } while (index < 0x80);

  *entries = index;

  DBG_LEAVE (); //EXIT //count_order
}

File: adt2unit/pas/count_order.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure count_order(var entries: Byte);

var
  index,
  index2: Byte;

begin
  _dbg_enter ({$I %FILE%}, 'count_order');

  index := 0;
  index2 := 0;

  Repeat
    If (songdata.pattern_order[index] <> $80) then
      begin
        If (songdata.pattern_order[index] > $80) then
          If (songdata.pattern_order[index]-$80 <> index2) then
            begin
              index := songdata.pattern_order[index]-$80;
              index2 := index;
            end
          else BREAK;
      end
    else BREAK;
    If (index < $80) then Inc(index);
  until (index > $7f);

  entries := index;

  _dbg_leave; //EXIT //count_order
end;
